Windows,client: extract embedded binaries faster

laszlocsomor opened the pull request:
When extracting embedded binaries, the client now
caches which directories it has already created
and won't attempt creating them again.

This saves some time on Windows: from 16.3 sec on
average down to 13.2 sec. (n=10 runs, always
starting Bazel with a new --output_user_root and
shutting down afterwards.)

On Linux I see only a marginal speedup, not
significant enough to claim credit for it. :)

See https://github.com/bazelbuild/bazel/issues/5444

laszlocsomor commented:
Based on https://github.com/bazelbuild/bazel/pull/5447.

cvcal commented on src/main/cpp/blaze.cc:
```diff
@@ -893,9 +893,11 @@ class ExtractBlazeZipProcessor : public PureZipExtractorProcessor {
-    if (!blaze_util::MakeDirectories(blaze_util::Dirname(path), 0777)) {
-      BAZEL_DIE(blaze_exit_code::INTERNAL_ERROR)
-          << "couldn't create '" << path << "': " << GetLastErrorString();
+    if (created_directories_.insert(path).second) {
```
It's not obvious why this is here, I would add a comment here explaining that you do this to prevent duplicate work for performance reasons. Otherwise, I could imagine someone accidentally removing this without testing the performance implications.

laszlocsomor replied:
Done.

cvcal commented on src/test/cpp/util/file_test.cc:
```diff
@@ -148,37 +148,31 @@ TEST(FileTest, TestMtimeHandling) {
-  ASSERT_TRUE(mtime.get()->GetIfInDistantFuture(tempdir, &actual));
-  ASSERT_FALSE(actual);
-
+  // Assert that a directory is always a good embedded embedded binary. (We do
```
embedded embedded

laszlocsomor replied:
Done.

cvcal commented on src/main/cpp/util/file_platform.h:
```diff
@@ -31,10 +31,11 @@ class IFileMtime {
+  // embedded binary that Bazel extracted from a previous run and it hasn't
+  // been tampered with (its mtime is unchanged), or it's a directory.
+  // Returns false otherwise, or if querying the information failed.
+  virtual bool IsValidEmbeddedBinary(const std::string &path) = 0;
```
Ok, I hear you on not wanting to move this to blaze_util_platform. Maybe you can make this name more general then, though? IsUntamperedFile() keeps it more neutral.

laszlocsomor replied:
I like that idea! Done.

laszlocsomor replied:
I also added a TODO with my own name to move the function.

cvcal commented on src/test/cpp/util/file_windows_test.cc:
```diff
@@ -34,6 +34,9 @@
 #error("This test should only be run on Windows")
 #endif  // !defined(_WIN32) && !defined(__CYGWIN__)
 
+#define _T(x) #x
```
These two macros look quite odd to me. I also don't see why you want to use __LINE__ as a way to distinguish between targets. You could just give a random number, no? Is there any point to use the line number from the file?

If you do need this, I don't think you need to split it into two definitions, the single one should be enough.

laszlocsomor replied:
>  I also don't see why you want to use \_\_LINE\_\_ as a way to distinguish between targets.

Because it is guaranteed to be unique among all other tests. For a while there was no cleanup code that deleted temp files that tests created. Now there is, at least in some tests (the test framework won't do it automatically), but I still use this practice to avoid tests clobbering each other's files.

> If you do need this, I don't think you need to split it into two definitions, the single one should be enough.

Indeed! Done.

laszlocsomor commented:
I addressed all comments, rebased this PR on top of `master`, and removed the diffbase (PR #5445) to make it cleaner.

Sorry for the confusion and for the confusing-looking commit history. _Now_ it should look cleaner.
